### ISSUE TITLE
Auto-implicit not evaluating expression

### ISSUE THREAD

**Issue by polendri** [idris-lang/Idris2#606]
Idris version: 0.2.1-2cd7350fb

I'm trying to write a parser (using `Text.Lexer` and `Text.Parser`) for use at compile-time to parse string literals into my regular expression type. The parser expression doesn't evaluate and so the type checking errors out, but there's no discernible reason why the expression wasn't evaluated. All functions in the module are total (via %default total).

I don't see why it would be specific to `Text.Parser`, but in trying to narrow things down, I couldn't find another function that reproduced the issue (no luck using an arbitrary `bar : String -> Maybe Char` function for example).

# Steps to Reproduce

See gist, which implements a parser of a single character: https://gist.github.com/polendri/fd8f0d2502bede194f874f4700dee1cf

The key portion is this (where `literalR` is a `Grammar  and `fooTokenMap` is a `TokenMap`):

```
tryParseFoo : String -> Maybe (Char)
tryParseFoo s = case parse literalR $ fst $ lex fooTokenMap s of
                     Right (c, []) => Just c
                     _ => Nothing

parseFoo : (s : String) -> {auto prf : IsJust (tryParseFoo s)} -> Char
parseFoo s {prf} with (tryParseFoo s)
  parseFoo s {prf=ItIsJust} | (Just c) = c
```

# Expected Behavior

```
> tryParseFoo "a"
Just 'a'
> parseFoo "a"
'a'
```

# Observed Behavior

```
> tryParseFoo "a"
Just 'a'
> parseFoo "a"
(interactive):1:1--1:13:Can't find an implementation for IsJust (case parse (terminal "Expected literal" (\x => case tok x of { Literal s => case unpack s of { [] => Nothing ; c :: _ => Just c } ; _ => Nothing })) (fst (lex [(non (oneOf "()"), Literal)] "a")) of { Right (c, []) => Just c ; _ => Nothing }) at:
1       parseFoo "a"
        ^^^^^^^^^^^^
```

Notably, I can copy-paste the expression from the error message into a definition in the module, and it evaluates to `Just 'a'` fine. So it's not clear to me why the compiler would not evaluate it down to `Just 'a'` and solve the auto-implicit?

It's not just a REPL issue, since attempting to call `parseFoo` in the module yields the same error at compile time:
```
test2 : Char
test2 = parseFoo "a"
```

**Comment by gallais** [comment]
Just in time to try my recent addition to logging: [`eval.stuck`](https://github.com/idris-lang/Idris2/commit/c17c6fc5220b72187d20a3f14895b8c3bdc6e179)

Adding `%logging eval.stuck 5` in your file, I get:

```idris
Loaded file Issue606.idr
Issue606> test
LOG eval.stuck:5: Stuck function: Text.Parser.Core.parse
LOG eval.stuck:5: Stuck function: Text.Parser.Core.doParse
LOG eval.stuck:5: Stuck function: Text.Parser.Core.terminal
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.lex
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.tokenise
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.non
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.<+>
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.reject
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.isJust
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.oneOf
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.pred
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.any
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.pred
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.tokenise
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.non
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.<+>
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.reject
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.isJust
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.oneOf
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.pred
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.any
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.pred
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.lex
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.tokenise
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.non
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.<+>
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.reject
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.isJust
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.oneOf
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.pred
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.any
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.pred
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.tokenise
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.non
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.<+>
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.reject
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.isJust
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.oneOf
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.pred
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.scan
LOG eval.stuck:5: Stuck function: Text.Lexer.any
LOG eval.stuck:5: Stuck function: Text.Lexer.Core.pred
Just 'a'
```

And indeed these functions are `export`ed but not `public`. That is to say that
their name & type are in scope when you import the module but their definition is
not available for reduction *at the type level*! This is done to ensure that users
of a library cannot possibly rely on the implementation details of a function and
get a breaking change if the library writer decides to perform a refactoring.

The REPL happily runs any expression you give it (as long as it can type it
without running non `public` code!) so the fact you are getting a `Just` back
is not enough for the expression to actually evaluate to a `Just` at the type level.

**Comment by polendri** [comment]
@gallais That's a handy piece of logging, thanks very much for the explanation!

I understand what's happening now, though I'm trying to work through the implications.

1. From an API perspective it seems like a good idea then that `Text.Lexer` and `Text.Parser` don't `public export` their contents, but effectively this makes them impossible to use for this purpose (to parse at the type level), correct? It seems to me that if a function similar to what I'm writing were defined _within_ the modules themselves, then it would work since the definitions are local. Would a feature request something like this solve my problem? (Happy to file that separately and contribute if you think it would)
    * Add `Text.Lexer.tryLex : TokenMap a -> String -> Maybe (List (TokenData a))`
    * Add `Text.Lexer.tLex : (tm : TokenMap a) -> (s : String) -> {auto prf : IsJust (tryLex tm s)} -> TokenData a` (needs a proper name)
    * Add `Text.Parser.tryParse : Grammar toc c ty -> List tok -> Maybe ty`
    * Add `Text.Parser.tParse : (g : Grammar toc c ty ) -> (ts : List tok) -> {auto prf : IsJust (tryParse g ts)} -> ty` (needs a proper name)
2. The error message doesn't provide any reason why the expression wasn't evaluated further, which leaves novices like me stumped. Could it conceivably do so, i.e. add something like a "Note: Definition of parse not available; definition of lex not available" message? If so, I'm not sure whether to file that as a feature request or whether there are bigger fish to fry these days than making slightly friendlier error messages.

**Comment by gallais** [comment]
> It seems to me that if a function similar to what I'm writing
  were defined within the modules themselves, then it would work
  since the definitions are local.

It would not. That function would get unfolded but then the evaluation
would get stuck on whatever internal representation got exposed.

This is tested [here](https://github.com/idris-lang/Idris2/tree/master/tests/idris2/basic045) where you can see in `Lib.idr` that `accMap` is public
but `accMapAux` is not.

If you look at the `expect` output when calling `accMap` from `Main.idr`,
you will see that `accMap` unfolds just fine but then `accMapAux` is reported
as stuck.



Basically, the parsing library is not designed to be used at the
type-level at compile time.

**Comment by polendri** [comment]
Right... if the definition is `export`, then it's not going to be evaluatable outside of that module no matter what I do. Enabling that would require things to be `public export` all the way down, which makes the APIs for `Text.Parser` and `Text.Lexer` too unstable. Seems to me the only solution is a 3rd party or custom parser library which accepts having a fully `public export` API.

Thanks again for the help. I won't clog the issue tracker with an error message request, hopefully for now this thread will help if anyone runs into a similar issue.
